File: src/core/Alerts/StockAlertContainer.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace core.Alerts
{
    public class StockAlertContainer
    {
        private record struct StockPositionMonitorKey(string description, string ticker, Guid userId);
        private ConcurrentDictionary<StockPositionMonitorKey, TriggeredAlert> _alerts =
            new ConcurrentDictionary<StockPositionMonitorKey, TriggeredAlert>();
        private const int MAX_RECENT_ALERTS = 20;
        
        private bool _manualRun = false;
        public void RequestManualRun() => _manualRun = true;
        public bool ManualRunRequested() => _manualRun;
        public void ManualRunCompleted() => _manualRun = false;

        public void Register(TriggeredAlert alert)
        {
            var key = new StockPositionMonitorKey(alert.identifier, alert.ticker, alert.userId);

            var firstAdd = !_alerts.ContainsKey(key);

            _alerts[key] = alert;

            if (firstAdd)
            {
                AddToRecent(alert);
            }
        }

        public void Deregister(string identifier, string ticker, Guid userId) =>
            _alerts.TryRemove(
                new StockPositionMonitorKey(identifier, ticker, userId),
                out _
            );
            

        private Dictionary<Guid, List<TriggeredAlert>> _recentlyTriggeredAlerts =
            new Dictionary<Guid, List<TriggeredAlert>>();

        public List<TriggeredAlert> GetRecentlyTriggeredAlerts(Guid userId) =>
            (_recentlyTriggeredAlerts.SingleOrDefault(u => u.Key == userId).Value ?? new List<TriggeredAlert>())
            .OrderByDescending(a => a.when)
            .ToList();

        public List<TriggeredAlert> GetAlerts(Guid userId) => 
            _alerts.Values.Where(m => m.userId == userId)
            .ToList();

        private void AddToRecent(TriggeredAlert triggeredAlert)
        {
            if (!_recentlyTriggeredAlerts.ContainsKey(triggeredAlert.userId))
            {
                _recentlyTriggeredAlerts[triggeredAlert.userId] = new List<TriggeredAlert>(MAX_RECENT_ALERTS);
            }

            var list = _recentlyTriggeredAlerts[triggeredAlert.userId];

            if (list.Count + 1 == MAX_RECENT_ALERTS)
            {
                list.RemoveAt(0);
            }

            list.Add(triggeredAlert);
        }

        public record struct AlertContainerMessage(string message, DateTimeOffset when);
        private int _messageLimit = 10;
        private List<AlertContainerMessage> _messages = new List<AlertContainerMessage>();
        public void AddNotice(string message)
        {
            _messages.Add(new AlertContainerMessage(message, DateTimeOffset.Now));
            if (_messages.Count > _messageLimit)
            {
                _messages.RemoveAt(0);
            }
        }

        public IEnumerable<AlertContainerMessage> GetMessages() =>
            _messages.OrderByDescending(m => m.when);

        private bool _listChecksCompleted;
        private bool _stopLossCheckCompleted;

        public void SetListCheckCompleted(bool completed) => _listChecksCompleted = completed;
        public void SetStopLossCheckCompleted(bool completed) => _stopLossCheckCompleted = completed;
        public bool ContainerReadyForNotifications() => _listChecksCompleted && _stopLossCheckCompleted;
    }
}

File: src/core.fs/Alerts/StockAlertContainer.fs
namespace core.fs.Alerts

    open System
    open System.Collections.Concurrent
    open core.Account
    open core.Shared
    open core.fs.Services.Analysis
    open core.fs.Services.GapAnalysis
    open core.fs.Shared
    open core.fs.Shared.Domain.Accounts

    type private StockPositionMonitorKey = {
        Ticker: Ticker
        UserId: UserId
    }
    
    [<Struct>]
    type AlertContainerMessage = {
        message: string
        ``when``: DateTimeOffset
    }
    
    [<Struct>]
    type AlertCheck = {
        ticker: Ticker
        listName: string
        user: UserState
    }
    
    type AlertType =
        | Negative = 0
        | Neutral = 1
        | Positive = 2
        
    module Constants =
        let MonitorTagPattern = "monitor:patterns"
        let MonitorNamePattern = "Patterns"
        let StopLossIdentifier = "Stop loss"
    
    
    [<Struct>]
    type TriggeredAlert =
        {
            identifier:string
            triggeredValue:decimal
            watchedValue:decimal
            ``when``:DateTimeOffset
            ticker:Ticker
            description:string
            sourceList:string
            userId:UserId
            alertType:AlertType
            valueFormat:ValueFormat
        }
        
        member this.Age() = (DateTimeOffset.UtcNow - this.``when``)
        
        static member GapUpAlert ticker sourceList (gap:Gap) ``when`` userId =
            {
                identifier = "Gap up"
                triggeredValue = gap.GapSizePct
                watchedValue = gap.GapSizePct
                ``when`` = ``when``
                ticker = ticker
                description = "Gap up"
                sourceList = sourceList
                userId = userId
                alertType = AlertType.Positive
                valueFormat = ValueFormat.Percentage 
            }
            
        static member StopPriceAlert ticker price stopPrice ``when`` userId =
            {
                identifier = Constants.StopLossIdentifier
                triggeredValue = price
                watchedValue = stopPrice
                ``when`` = ``when``
                ticker = ticker
                description = "Stop price"
                sourceList = "Stop price"
                userId = userId
                alertType = AlertType.Negative
                valueFormat = ValueFormat.Currency
            }
            
        static member PatternAlert (pattern:Pattern) ticker sourceList ``when`` userId =
            {
                identifier = pattern.name
                triggeredValue = pattern.value
                watchedValue = pattern.value
                ``when`` = ``when``
                ticker = ticker
                description = pattern.description
                sourceList = sourceList
                userId = userId
                alertType = AlertType.Neutral
                valueFormat = pattern.valueFormat
            }
        
    type StockAlertContainer() =
        let alerts = ConcurrentDictionary<StockPositionMonitorKey, TriggeredAlert>()
        let recentlyTriggered = ConcurrentDictionary<UserId, List<TriggeredAlert>>()
        let notices = System.Collections.Generic.List<AlertContainerMessage>(capacity=10)
        
        let mutable manualRun = false
        let mutable listChecksCompleted = false
        let mutable stopLossCheckCompleted = false
        
        let addToRecent (alert:TriggeredAlert) =
            
            match recentlyTriggered.TryAdd(key=alert.userId, value=[alert]) with
            | true -> ()
            | false ->
                let list = recentlyTriggered[alert.userId]
                
                let newList =
                    match list.Length with
                    | 20 -> list |> List.tail
                    | _ -> list
                
                recentlyTriggered[alert.userId] <- (newList @ [alert])
        
        member _.RequestManualRun() = manualRun <- true;
        member _.ManualRunRequested() = manualRun
        member _.ManualRunCompleted() = manualRun <- false
        
        
        member _.Register (alert:TriggeredAlert) =
            let key = { Ticker = alert.ticker; UserId = alert.userId }
            
            match alerts.TryAdd(key, alert) with
            | true -> alert |> addToRecent
            | false -> alerts[key] = alert |> ignore
            
        member _.DeregisterAlert (alert:TriggeredAlert) =
            
            { Ticker = alert.ticker; UserId = alert.userId }
            |> alerts.TryRemove |> ignore
            
        member _.Deregister _ ticker userId =
            { Ticker = ticker; UserId = userId }
            |> alerts.TryRemove |> ignore
            
        member _.DeregisterStopPriceAlert ticker userId =
            { Ticker = ticker; UserId = userId }
            |> alerts.TryRemove |> ignore
            
        member _.GetRecentlyTriggered userId =
            let list =
                match recentlyTriggered.TryGetValue(userId) with
                | true, alerts -> alerts
                | false, _ -> []
            
            list |> List.sortByDescending (fun x -> x.``when``)
            
        member _.GetAlerts userId =
            alerts.Values
            |> Seq.filter (fun alert -> alert.userId = userId)
            |> Seq.sortBy (fun x -> x.sourceList,x.ticker.Value)
            
        member _.AddNotice message =
            if notices.Count = 10 then
                notices.RemoveAt(0)
            
            { message = message; ``when`` = DateTimeOffset.UtcNow }
            |> notices.Add
            
        member _.GetNotices () =
            notices |> Seq.sortByDescending (fun x -> x.``when``)

        member _.SetListCheckCompleted completed = listChecksCompleted <- completed
        member _.SetStopLossCheckCompleted completed = stopLossCheckCompleted <- completed
        member _.ContainerReadyForNotifications() = listChecksCompleted && stopLossCheckCompleted
        member this.ClearStopLossAlert() =
            alerts.Values
            |> Seq.filter (fun x -> x.identifier = Constants.StopLossIdentifier)
            |> Seq.map (fun x -> { Ticker = x.ticker; UserId = x.userId })
            |> Seq.iter (fun x -> alerts.TryRemove(x) |> ignore)
            